tests/integration/test_asr_gradient_step_and_eval.py

```diff
@@ -287,81 +232,26 @@ def test_stft_conv_training(self):
-            'window_size': 0.02,
-            'n_fft': 512,
-            'dither': 1e-05,
-            'window': 'hann',
-            'sample_rate': 16000,
-            'normalize': 'per_feature',
-            'window_stride': 0.01,
-            'stft_conv': True,
-        }
-        preprocessing = nemo_asr.AudioToMelSpectrogramPreprocessor(**pre_process_params)
-        encoder = nemo_asr.JasperEncoder(
-            jasper=cfg['encoder']['jasper'],
-            activation=cfg['encoder']['activation'],
-            feat_in=cfg['input']['train']['features'],
-        )
-        connector = nemo_asr.JasperRNNConnector(
-            in_channels=cfg['encoder']['jasper'][-1]['filters'], out_channels=cfg['decoder']['hidden_size'],
+        self.nf.train(
+            [loss], callbacks=[callback], optimizer="sgd", optimization_params={"max_steps": 3, "lr": 0.003},
         )
-        decoder = nemo.backends.pytorch.common.DecoderRNN(
-            voc_size=len(self.labels),
-            bos_id=0,
-            hidden_size=cfg['decoder']['hidden_size'],
-            attention_method=cfg['decoder']['attention_method'],
-            attention_type=cfg['decoder']['attention_type'],
-            in_dropout=cfg['decoder']['in_dropout'],
-            gru_dropout=cfg['decoder']['gru_dropout'],
-            attn_dropout=cfg['decoder']['attn_dropout'],
-            teacher_forcing=cfg['decoder']['teacher_forcing'],
-            curriculum_learning=cfg['decoder']['curriculum_learning'],
-            rnn_type=cfg['decoder']['rnn_type'],
-            n_layers=cfg['decoder']['n_layers'],
-            tie_emb_out_weights=cfg['decoder']['tie_emb_out_weights'],
-        )
-        loss = nemo.backends.pytorch.common.SequenceLoss()
-
-        # DAG
-        audio_signal, a_sig_length, transcripts, transcript_len = dl()
-        processed_signal, p_length = preprocessing(input_signal=audio_signal, length=a_sig_length)
-        encoded, encoded_len = encoder(audio_signal=processed_signal, length=p_length)
-        encoded = connector(tensor=encoded)
-        log_probs, _ = decoder(targets=transcripts, encoder_outputs=encoded)
-        loss = loss(log_probs=log_probs, targets=transcripts)
+        self.nf.reset_trainer()
 
-        # Train
-        callback = nemo.core.SimpleLossLoggerCallback(
-            tensors=[loss], print_func=lambda x: logging.info(str(x[0].item()))
-        )
-        # Instantiate an optimizer to perform `train` action
-        optimizer = nemo.backends.pytorch.actions.PtActions()
-        optimizer.train(
-            [loss], callbacks=[callback], optimizer="sgd", optimization_params={"num_epochs": 10, "lr": 0.0003},
-        )
+        # Assert that training loss went down
+        assert loss_list[-1] < loss_list[0]
 
-    @pytest.mark.unclassified
+    @pytest.mark.integration
+    @pytest.mark.skipduringci
     def test_jasper_evaluation(self):
+        # Note this test still has no asserts, but rather checks that the current eval path works
```

Comment:
Ok, so when do we want this test to be executed?

During nightly build? Weekly build?

How about introducing a new marker: convergence?

I define Convergence Tests as a subclass of Acceptance tests, aiming at making sure that models that are crucial for us from business perspective (e.g. JASPER as our flagship model) or scientific perspective (e.g. in our tutorials we report scores for TRADE on MultiWOZ, so we need to make sure that it still works) converge.

@blisc What do you think about that?

Reply:
This is not a convergence test as in weights are random. While this test creates a Jasper graph, it mainly checks the following which are not currently covered by any other pytest test:

- EvaluatorCallback
- NerualModuleFactory.eval

To me, this would be useful for users to test after installing NeMo but prior to model training to verify that their installation worked as intended. I have marked it as skipduringci as it is redundant with our Jasper Jenkins CI tests, and thus we do not need to test this as part of our testing setup.

I am unsure what the difference between an integration and an acceptance test is. To me, I am integrating the callback with our eval function to ensure that it works. Do you think this test is an acceptance test instead?

Reply:
1) eval is tested in system tests
2) I have a "gut feeling" that during those tests we should not call evaluation and rely only on training loss - faster, the same result (we are not testing the generalization, right?)

So what you are saying clearly indicates that NeMo lacks a test suite for callbacks, maybe as part of Callbacks redesign/refactor we will take care of it, for now let's leave it the way you implemented it, those are good changes!! 👍